### PR TITLE
Mkaufman mongodb

### PR DESCRIPTION
**Pull request by mike-kaufman**
This is very initial support for instrumenting mongoDB driver.

### CODE REVIEW

File: src/glimpse.agent/inspectors/mongodb-proxy.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                        .then(
+                            function insertSuccess(r) {
+                                sendAfterExecuteMessage(broker, startTime);
+                                resolve(r);
+                            },
+                            function insertFailed(err) {
+                                sendAfterExecuteMessage(broker, startTime, err);  // TODO:  what message do we send in the event of an error??
+                                reject(err);
+                            });
+                });
```
Comment (avanderhoorn) [inline src/glimpse.agent/inspectors/mongodb-proxy.js]: Wonder if the above will turn into a common pattern.

File: src/glimpse.agent/inspectors/mongodb-proxy.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            }
+        }
+
+        /**
+         *  sendBeforeExecuteMessage 
+         */
+        var sendBeforeExecuteMessage = function sendBeforeExecuteMessage(broker, startTime) {
```
Comment (avanderhoorn) [inline src/glimpse.agent/inspectors/mongodb-proxy.js]: Does `sendBeforeExecuteMessage` & `sendAfterExecuteMessage` need to be defined inside the `onInit` function? I think the other proxies follow a pattern where they are outside of the init.

File: src/glimpse.agent/inspectors/mongodb-proxy.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+            sendBeforeExecuteMessage(broker, startTime);
+
+            if (originalCallback) {
+                // a callback was passed in
+                var callbackWrapper = function callbackWrapper(err, r) {
+                    sendAfterExecuteMessage(broker, startTime, err);
+                    originalCallback.apply(this, arguments);
+                }
+                realInsertMany.call(this, docs, options, callbackWrapper);
```
Comment (avanderhoorn) [inline src/glimpse.agent/inspectors/mongodb-proxy.js]: Wonder if this is this is better to do as an `apply` and replace the element `2` in the `arguments`?

File: src/glimpse.agent/inspectors/mongodb-proxy.js (hunk 1 above)
Comment (mike-kaufman) [inline src/glimpse.agent/inspectors/mongodb-proxy.js]: Yeah, it probably will become common.  Most of the mongoDB APIs either take a callback, and if no callback is specified, return a promise.  I'm going to work on instrumenting the query APIs today, so I'll refactor as necessary.

File: src/glimpse.agent/inspectors/mongodb-proxy.js (hunk 2 above)
Comment (mike-kaufman) [inline src/glimpse.agent/inspectors/mongodb-proxy.js]: fixed

File: src/glimpse.agent/inspectors/mongodb-proxy.js (hunk 3 above)
Comment (mike-kaufman) [inline src/glimpse.agent/inspectors/mongodb-proxy.js]: I went back and forth on this.  The weirdness with using apply is identifying where in the args the callback is - as if no options are specified, the callback is the second arg.  I'll fix this though to use apply.

**Comment by mike-kaufman** [comment]
Let me know if there are any other comments.  If not, I'll merge this in.

**Pull request merged**